src/qunit.cpp

```diff
@@ -1051,35 +1067,75 @@ void QUnit::INC(bitCapInt toMod, bitLenInt start, bitLenInt length)
     shards[start].unit->INC(toMod, shards[start].mapped, length);
 }
 
+bool QUnit::CArithmeticOptimize(bitLenInt start, bitLenInt length, bitLenInt* controls, bitLenInt controlLen, std::vector<bitLenInt>* controlVec) {
+    for (auto i = 0; i < controlLen; i++) {
+        // If any control has a cached zero probability, this gate will do nothing, and we can avoid basically all
+        // overhead.
+        if (!shards[controls[i]].isProbDirty && (Prob(controls[i]) < min_norm)) {
+            return true;
+        }
+    }
+
+    // Otherwise, we have to entangle the register.
+    EntangleRange(start, length);
+
+    controlVec->resize(controlLen);
+    std::copy(controls, controls + controlLen, controlVec->begin());
```

Comment:
Is it necessary to perform the copy operation when it's still possible there's no work to perform?

Reply:
By the time we've made it this deep, we've should have already checked that there is at least one control bit passed in. Through this method, it's possible we weed them all out as immaterial.

So, are you asking if we could save the overhead of the copy? Because, I suspect it's really nothing, compared to the gate itself, usually.

Reply:
Benn, rather than ending the review in the form of a question, my response is:

Yes, though the copy is pointer-safe, and the overhead is negligible compared to the gate itself.

Reply:
(Hypothetically, no, I can't think of way to decouple these aspects, without significant redundancy for the loops.)